File: backend/app/models/package.py
import uuid
from datetime import datetime

from app.extensions import db
from app.utils.datetime_format import utc_isoformat


class Package(db.Model):
    __tablename__ = "packages"

    id = db.Column(db.UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tracking_number = db.Column(db.String(30), unique=True, nullable=False, index=True)
    customer_id = db.Column(db.UUID(as_uuid=True), db.ForeignKey("users.id"), nullable=False)

    carrier_tracking = db.Column(db.String(100))
    label_name = db.Column(db.String(255))
    label_boss_id = db.Column(db.String(20))
    shipper = db.Column(db.String(30))
    actual_weight_lbs = db.Column(db.Numeric(8, 2))
    billable_weight_lbs = db.Column(db.Integer)
    shipping_cost_usd = db.Column(db.Numeric(10, 2))
    estimated_freight_jmd = db.Column(db.Numeric(12, 2))
    duties_jmd = db.Column(db.Numeric(12, 2))
    handling_jmd = db.Column(db.Numeric(12, 2))
    other_fees_jmd = db.Column(db.Numeric(12, 2))
    total_due_jmd = db.Column(db.Numeric(12, 2))
    billing_status = db.Column(db.String(20), nullable=False, default="pending")
    invoice_status = db.Column(db.String(20), nullable=False, default="pending")
    invoice_object_key = db.Column(db.String(500))
    declared_value_usd = db.Column(db.Numeric(10, 2))
    invoice_requested_at = db.Column(db.DateTime)
    invoice_requested_via = db.Column(db.String(20))
    invoice_request_note = db.Column(db.String(500))
    invoice_received_at = db.Column(db.DateTime)
    delivery_address_id = db.Column(db.UUID(as_uuid=True), db.ForeignKey("delivery_addresses.id"))
    rate_tier_label = db.Column(db.String(50))

    status = db.Column(db.String(50), nullable=False, default="received")
    receive_batch_id = db.Column(
        db.UUID(as_uuid=True), db.ForeignKey("receive_batches.id"), nullable=True, index=True
    )
    shipment_id = db.Column(db.UUID(as_uuid=True), db.ForeignKey("shipments.id"), nullable=True, index=True)
    label_printed_at = db.Column(db.DateTime, nullable=True)
    received_at = db.Column(db.DateTime, default=datetime.utcnow, nullable=False)
    created_at = db.Column(db.DateTime, default=datetime.utcnow, nullable=False)
    updated_at = db.Column(
        db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )

    customer = db.relationship("User", backref="packages")
    receive_batch = db.relationship("ReceiveBatch", back_populates="packages")
    shipment = db.relationship("Shipment", back_populates="packages")
    delivery_address = db.relationship("DeliveryAddress", backref="packages")
    photos = db.relationship("PackagePhoto", backref="package", lazy=True, cascade="all, delete-orphan")
    events = db.relationship(
        "PackageEvent",
        backref="package",
        lazy=True,
        order_by="PackageEvent.created_at",
        cascade="all, delete-orphan",
    )

    def to_dict(self, include_events: bool = False, include_photos: bool = False) -> dict:
        from app.constants import BILLING_STATUS_LABELS, INVOICE_STATUS_LABELS, SHIPPER_LABELS, STATUS_LABELS
        from app.services.image_upload_service import resolve_stored_url

        data = {
            "id": str(self.id),
            "tracking_number": self.tracking_number,
            "status": self.status,
            "status_label": STATUS_LABELS.get(self.status, self.status),
            "carrier_tracking": self.carrier_tracking,
            "label_name": self.label_name,
            "label_boss_id": self.label_boss_id,
            "is_unidentified": self.status == "unidentified",
            "shipper": self.shipper,
            "shipper_label": SHIPPER_LABELS.get(self.shipper, self.shipper) if self.shipper else None,
            "actual_weight_lbs": float(self.actual_weight_lbs) if self.actual_weight_lbs else None,
            "billable_weight_lbs": self.billable_weight_lbs,
            "estimated_freight_jmd": float(self.estimated_freight_jmd)
            if self.estimated_freight_jmd is not None
            else None,
            "duties_jmd": float(self.duties_jmd) if self.duties_jmd is not None else None,
            "handling_jmd": float(self.handling_jmd) if self.handling_jmd is not None else None,
            "other_fees_jmd": float(self.other_fees_jmd) if self.other_fees_jmd is not None else None,
            "total_due_jmd": float(self.total_due_jmd) if self.total_due_jmd is not None else None,
            "currency": "JMD",
            "billing_status": self.billing_status,
            "billing_status_label": BILLING_STATUS_LABELS.get(self.billing_status, self.billing_status),
            "invoice_status": self.invoice_status,
            "invoice_status_label": INVOICE_STATUS_LABELS.get(self.invoice_status, self.invoice_status),
            "invoice_object_key": self.invoice_object_key,
            "invoice_url": resolve_stored_url(self.invoice_object_key) if self.invoice_object_key else None,
            "declared_value_usd": float(self.declared_value_usd) if self.declared_value_usd else None,
            "invoice_requested_at": utc_isoformat(self.invoice_requested_at),
            "invoice_requested_via": self.invoice_requested_via,
            "invoice_request_note": self.invoice_request_note,
            "invoice_received_at": utc_isoformat(self.invoice_received_at),
            "delivery_address_id": str(self.delivery_address_id) if self.delivery_address_id else None,
            "delivery_address": self.delivery_address.to_dict() if self.delivery_address else None,
            "rate_tier_label": self.rate_tier_label,
            "label_printed_at": utc_isoformat(self.label_printed_at),
            "received_at": utc_isoformat(self.received_at),
            "receive_batch_id": str(self.receive_batch_id) if self.receive_batch_id else None,
            "shipment_id": str(self.shipment_id) if self.shipment_id else None,
            "created_at": utc_isoformat(self.created_at),
        }
        if self.receive_batch:
            data["receive_batch"] = {
                "id": str(self.receive_batch.id),
                "batch_code": self.receive_batch.batch_code,
                "reference": self.receive_batch.reference,
                "receive_date": self.receive_batch.receive_date.isoformat(),
                "status": self.receive_batch.status,
            }
        else:
            data["receive_batch"] = None
        if self.shipment:
            data["shipment"] = {
                "id": str(self.shipment.id),
                "reference": self.shipment.reference,
                "departure_date": self.shipment.departure_date.isoformat(),
                "status": self.shipment.status,
            }
        else:
            data["shipment"] = None
        if include_events:
            data["events"] = [e.to_dict() for e in self.events]
        if include_photos:
            data["photos"] = [p.to_dict() for p in self.photos]
        return data


class PackageEvent(db.Model):
    __tablename__ = "package_events"

    id = db.Column(db.UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    package_id = db.Column(db.UUID(as_uuid=True), db.ForeignKey("packages.id"), nullable=False)
    status = db.Column(db.String(50), nullable=False)
    note = db.Column(db.String(255))
    created_at = db.Column(db.DateTime, default=datetime.utcnow, nullable=False)

    def to_dict(self) -> dict:
        from app.constants import STATUS_LABELS

        return {
            "id": str(self.id),
            "status": self.status,
            "status_label": STATUS_LABELS.get(self.status, self.status),
            "note": self.note,
            "created_at": utc_isoformat(self.created_at),
        }


class PackagePhoto(db.Model):
    __tablename__ = "package_photos"

    id = db.Column(db.UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    package_id = db.Column(db.UUID(as_uuid=True), db.ForeignKey("packages.id"), nullable=False)
    r2_object_key = db.Column(db.String(500), nullable=False)
    created_at = db.Column(db.DateTime, default=datetime.utcnow, nullable=False)

    def to_dict(self) -> dict:
        from app.services.image_upload_service import resolve_stored_url

        return {
            "id": str(self.id),
            "object_key": self.r2_object_key,
            "url": resolve_stored_url(self.r2_object_key),
            "created_at": utc_isoformat(self.created_at),
        }

File: server/src/lib/serializers/package.ts
import {
  BILLING_STATUS_LABELS,
  INVOICE_STATUS_LABELS,
  SHIPPER_LABELS,
  STATUS_LABELS,
  UNIDENTIFIED_HOLDER_SHIPPING_ID,
} from '../../constants.js'
import type {
  deliveryAddresses,
  packageEvents,
  packagePhotos,
  packages,
  receiveBatches,
  shipments,
  users,
} from '../../db/schema/index.js'
import { utcIsoformat } from '../dates.js'
import { deliveryAddressToDict } from './deliveryAddress.js'
import { userFullName } from './user.js'
import { resolveStoredUrl } from '../../services/imageUploadService.js'

type PackageRow = typeof packages.$inferSelect
type PackageEventRow = typeof packageEvents.$inferSelect
type PackagePhotoRow = typeof packagePhotos.$inferSelect
type DeliveryAddressRow = typeof deliveryAddresses.$inferSelect
type ReceiveBatchRow = typeof receiveBatches.$inferSelect
type ShipmentRow = typeof shipments.$inferSelect
type UserRow = typeof users.$inferSelect

export function packageEventToDict(event: PackageEventRow): Record<string, unknown> {
  return {
    id: event.id,
    status: event.status,
    status_label: STATUS_LABELS[event.status] ?? event.status,
    note: event.note,
    created_at: utcIsoformat(event.createdAt),
  }
}

export function packagePhotoToDict(photo: PackagePhotoRow): Record<string, unknown> {
  return {
    id: photo.id,
    object_key: photo.r2ObjectKey,
    url: resolveStoredUrl(photo.r2ObjectKey),
    created_at: utcIsoformat(photo.createdAt),
  }
}

export function packageToDict(
  pkg: PackageRow,
  opts: {
    includeEvents?: boolean
    includePhotos?: boolean
    events?: PackageEventRow[]
    photos?: PackagePhotoRow[]
    deliveryAddress?: DeliveryAddressRow | null
    receiveBatch?: ReceiveBatchRow | null
    shipment?: ShipmentRow | null
  } = {},
): Record<string, unknown> {
  const data: Record<string, unknown> = {
    id: pkg.id,
    tracking_number: pkg.trackingNumber,
    status: pkg.status,
    status_label: STATUS_LABELS[pkg.status] ?? pkg.status,
    carrier_tracking: pkg.carrierTracking,
    item_description: pkg.itemDescription,
    label_name: pkg.labelName,
    label_boss_id: pkg.labelBossId,
    is_unidentified: pkg.status === 'unidentified',
    shipper: pkg.shipper,
    shipper_label: pkg.shipper ? (SHIPPER_LABELS[pkg.shipper] ?? pkg.shipper) : null,
    actual_weight_lbs: pkg.actualWeightLbs != null ? parseFloat(pkg.actualWeightLbs) : null,
    billable_weight_lbs: pkg.billableWeightLbs,
    estimated_freight_jmd:
      pkg.estimatedFreightJmd != null ? parseFloat(pkg.estimatedFreightJmd) : null,
    duties_jmd: pkg.dutiesJmd != null ? parseFloat(pkg.dutiesJmd) : null,
    handling_jmd: pkg.handlingJmd != null ? parseFloat(pkg.handlingJmd) : null,
    other_fees_jmd: pkg.otherFeesJmd != null ? parseFloat(pkg.otherFeesJmd) : null,
    total_due_jmd: pkg.totalDueJmd != null ? parseFloat(pkg.totalDueJmd) : null,
    currency: 'JMD',
    billing_status: pkg.billingStatus,
    billing_status_label: BILLING_STATUS_LABELS[pkg.billingStatus] ?? pkg.billingStatus,
    invoice_status: pkg.invoiceStatus,
    invoice_status_label: INVOICE_STATUS_LABELS[pkg.invoiceStatus] ?? pkg.invoiceStatus,
    invoice_object_key: pkg.invoiceObjectKey,
    invoice_url: pkg.invoiceObjectKey ? resolveStoredUrl(pkg.invoiceObjectKey) : null,
    declared_value_usd: pkg.declaredValueUsd != null ? parseFloat(pkg.declaredValueUsd) : null,
    invoice_requested_at: utcIsoformat(pkg.invoiceRequestedAt),
    invoice_requested_via: pkg.invoiceRequestedVia,
    invoice_request_note: pkg.invoiceRequestNote,
    invoice_received_at: utcIsoformat(pkg.invoiceReceivedAt),
    delivery_address_id: pkg.deliveryAddressId,
    delivery_address: opts.deliveryAddress ? deliveryAddressToDict(opts.deliveryAddress) : null,
    rate_tier_label: pkg.rateTierLabel,
    label_printed_at: utcIsoformat(pkg.labelPrintedAt),
    received_at: utcIsoformat(pkg.receivedAt),
    receive_batch_id: pkg.receiveBatchId,
    shipment_id: pkg.shipmentId,
    created_at: utcIsoformat(pkg.createdAt),
  }

  if (opts.receiveBatch) {
    data.receive_batch = {
      id: opts.receiveBatch.id,
      batch_code: opts.receiveBatch.batchCode,
      reference: opts.receiveBatch.reference,
      receive_date: opts.receiveBatch.receiveDate,
      status: opts.receiveBatch.status,
    }
  } else {
    data.receive_batch = null
  }

  if (opts.shipment) {
    data.shipment = {
      id: opts.shipment.id,
      reference: opts.shipment.reference,
      departure_date: opts.shipment.departureDate,
      status: opts.shipment.status,
    }
  } else {
    data.shipment = null
  }

  if (opts.includeEvents && opts.events) {
    data.events = opts.events.map(packageEventToDict)
  }
  if (opts.includePhotos && opts.photos) {
    data.photos = opts.photos.map(packagePhotoToDict)
  }

  return data
}

export function warehousePackageListToDict(
  pkg: PackageRow,
  opts: {
    customer?: UserRow | null
    shipment?: ShipmentRow | null
  } = {},
): Record<string, unknown> {
  const customer = opts.customer
  const customerData =
    customer && customer.shippingId !== UNIDENTIFIED_HOLDER_SHIPPING_ID
      ? {
          id: customer.id,
          full_name: userFullName(customer),
          shipping_id: customer.shippingId,
        }
      : null

  const shipmentData = opts.shipment
    ? {
        id: opts.shipment.id,
        reference: opts.shipment.reference,
        departure_date: opts.shipment.departureDate,
        status: opts.shipment.status,
      }
    : null

  return {
    id: pkg.id,
    tracking_number: pkg.trackingNumber,
    carrier_tracking: pkg.carrierTracking,
    item_description: pkg.itemDescription,
    status: pkg.status,
    status_label: STATUS_LABELS[pkg.status] ?? pkg.status,
    invoice_status: pkg.invoiceStatus,
    invoice_status_label: INVOICE_STATUS_LABELS[pkg.invoiceStatus] ?? pkg.invoiceStatus,
    invoice_url: pkg.invoiceObjectKey ? resolveStoredUrl(pkg.invoiceObjectKey) : null,
    billing_status: pkg.billingStatus,
    billing_status_label: BILLING_STATUS_LABELS[pkg.billingStatus] ?? pkg.billingStatus,
    total_due_jmd: pkg.totalDueJmd != null ? parseFloat(pkg.totalDueJmd) : null,
    actual_weight_lbs: pkg.actualWeightLbs != null ? parseFloat(pkg.actualWeightLbs) : null,
    billable_weight_lbs: pkg.billableWeightLbs,
    received_at: utcIsoformat(pkg.receivedAt),
    customer: customerData,
    shipment: shipmentData,
  }
}

export function warehousePackageToDict(
  pkg: PackageRow,
  opts: {
    customer?: UserRow | null
    deliveryAddress?: DeliveryAddressRow | null
    receiveBatch?: ReceiveBatchRow | null
    shipment?: ShipmentRow | null
    events?: PackageEventRow[]
    photos?: PackagePhotoRow[]
  } = {},
): Record<string, unknown> {
  const data = packageToDict(pkg, {
    includeEvents: true,
    includePhotos: true,
    events: opts.events,
    photos: opts.photos,
    deliveryAddress: opts.deliveryAddress,
    receiveBatch: opts.receiveBatch,
    shipment: opts.shipment,
  })

  const customer = opts.customer
  if (customer && customer.shippingId !== UNIDENTIFIED_HOLDER_SHIPPING_ID) {
    data.customer = {
      id: customer.id,
      full_name: userFullName(customer),
      email: customer.email,
      shipping_id: customer.shippingId,
      parish: customer.parish,
    }
  } else {
    data.customer = null
  }
  return data
}
